aiohttp/helpers.py

```diff
@@ -279,6 +329,246 @@ def basicauth_from_netrc(netrc_obj: Optional[netrc.netrc], host: str) -> BasicAu
+            raise client_exceptions.ClientError("Challenge is missing nonce")
+
+        realm: str = self.ctx.challenge.get("realm", "")
+        nonce: str = self.ctx.challenge.get("nonce", "")
+        qop_raw: str = self.ctx.challenge.get("qop", "")
+        algorithm: str = self.ctx.challenge.get("algorithm", "MD5").upper()
+        opaque: str = self.ctx.challenge.get("opaque", "")
+
+        qop: str = ""
+        if qop_raw:
+            qop_list = [q.strip() for q in qop_raw.split(",") if q.strip()]
+            valid_qops = {"auth", "auth-int"}.intersection(qop_list)
+            if not valid_qops:
+                raise client_exceptions.ClientError(
+                    f"Unsupported qop value(s): {qop_raw}"
+                )
+
+            qop = "auth-int" if "auth-int" in valid_qops else "auth"
+
+        if algorithm not in DigestFunctions:
+            return ""
+        hash_fn: Final = DigestFunctions[algorithm]
+
+        def H(x: str) -> str:
+            return hash_fn(x.encode()).hexdigest()
+
+        def KD(s: str, d: str) -> str:
+            return H(f"{s}:{d}")
+
+        path = URL(url).path_qs
+        A1 = f"{self.login}:{realm}:{self.password}"
+        A2 = f"{method.upper()}:{path}"
+        if qop == "auth-int":
+            if isinstance(body, bytes):
+                entity_str = body.decode("utf-8", errors="replace")
+            elif isinstance(body, str):
+                entity_str = body
+            else:
+                entity_str = ""
+            entity_hash = H(entity_str)
+            A2 = f"{A2}:{entity_hash}"
+
+        HA1 = H(A1)
+        HA2 = H(A2)
+
+        if nonce == self.ctx.last_nonce:
+            self.ctx.nonce_count += 1
+        else:
+            self.ctx.nonce_count = 1
+
+        self.ctx.last_nonce = nonce
+
+        ncvalue = f"{self.ctx.nonce_count:08x}"
+
+        # cnonce is just a random string generated by the client.
+        cnonce_data = "".join(
+            [
+                str(self.ctx.nonce_count),
+                nonce,
+                time.ctime(),
+                os.urandom(8).decode(errors="ignore"),
+            ]
+        ).encode()
+        cnonce = hashlib.sha1(cnonce_data).hexdigest()[:16]
+
+        if algorithm == "MD5-SESS":
+            HA1 = H(f"{HA1}:{nonce}:{cnonce}")
+
+        if qop:
+            noncebit = ":".join([nonce, ncvalue, cnonce, qop, HA2])
+            response_digest = KD(HA1, noncebit)
+        else:
+            response_digest = KD(HA1, f"{nonce}:{HA2}")
+
+        pairs = [
+            f'username="{self.login}"',
+            f'realm="{realm}"',
+            f'nonce="{nonce}"',
+            f'uri="{path}"',
+            f'response="{response_digest}"',
+            f'algorithm="{algorithm}"',
+        ]
+        if opaque:
+            pairs.append(f'opaque="{opaque}"')
+        if qop:
+            pairs.append(f'qop="{qop}"')
+            pairs.append(f"nc={ncvalue}")
+            pairs.append(f'cnonce="{cnonce}"')
```

Comment:
These need to be check for possible injection attacks in the event any of this could be user provided